tools/ci.sh

```diff
@@ -4,8 +4,11 @@ export RUST_BACKTRACE=1
 export CARGO_INCREMENTAL=0
 export CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse
 
-cargo build -p lalrpop
-cargo test --all --all-features
+cargo build --bin lalrpop --features pico-args
+# build with minimal amount of features to make sure dependencies can build too.
+cargo hack build --workspace --each-feature --optional-deps
```

Comment:
Since to run the tests you need to have compiled the corresponding project, is this redundant?(Maybe we only need `cargo hack test`)

Reply:
As far as I am aware, there is no guaranty that the binary (`cargo build -p lalrpop --features pico-args`) will be built before the builds that require it (`cargo build --no-default-features --features test` is one of them)

Reply:
I don't think I'm confident enough on the first point to speculate beyond just saying I removed this line locally and it still worked for me. But you probably understand this point better than I do.

For the second point, at least in running the script locally and in looking at the logs, `cargo hack test` does seem to recompile on some of the sets of features. In general, I think it's a fair concern that the ci goes from 6 min https://github.com/lalrpop/lalrpop/actions/runs/4833729044/jobs/8614076121 to 12 min https://github.com/lalrpop/lalrpop/actions/runs/4834066012/jobs/8614838843?pr=786 . I'm happy to try to mitigate this in a different pr if that's easier.

Reply:
I did the change locally.
I am looking at documentation updates now.

Reply:
The doubling in CI time is because instead of testing 1 set of features (all enabled) it's building a lot more combinasion (even more when powerset is enabled).
But IMHO, exhaustiveness is the point of having CI as a strong PR gate.
CI may benefit from running build & tests in parallel pipelines. 

I can make something similar to this: https://github.com/rp-rs/rp-hal/pull/588
The downside being that it's not using a shell script, so it's sell trivial to run the same tests than the pipeline locally before pushing.